README.md

# OAuth

## About

Oauth consumer implementation.


CHANGELOG.md

### 2.1.1 (2018-12-16)

* Upgrade PHPUnit to 6.0

### 2.1.0 (2018-02-14)

* Use guzzle mock in test case
* Updated http dependency
* Removed data dependency

### 2.0.0 (2016-12-17)

* Set php 5.6 to min requirement
* Update psx data to 2.0

### 1.0.0 (2016-05-08)

* Initial release


tests/ConsumerTest.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth\Tests;

use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use PSX\Http\Authentication;
use PSX\Http\Client\Client;
use PSX\Http\Client\GetRequest;
use PSX\Oauth\Data;
use PSX\Oauth\Consumer;
use PSX\Uri\Url;

/**
 * ConsumerTest
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 */
class ConsumerTest extends TestCase
{
    const CONSUMER_KEY      = 'dpf43f3p2l4k3l03';
    const CONSUMER_SECRET   = 'kd94hf93k423kf44';

    const TMP_TOKEN         = 'hh5s93j4hdidpola';
    const TMP_TOKEN_SECRET  = 'hdhd0244k9j7ao03';
    const VERIFIER          = 'hfdp7dh39dks9884';
    const TOKEN             = 'nnch734d00sl2jdk';
    const TOKEN_SECRET      = 'pfkkdhi9sl3r4s00';

    public function testFlow()
    {
        $tmpToken       = self::TMP_TOKEN;
        $tmpTokenSecret = self::TMP_TOKEN_SECRET;
        $token          = self::TOKEN;
        $tokenSecret    = self::TOKEN_SECRET;

        $mock = new MockHandler([
            new Response(200, ['Content-Type' => 'application/x-www-form-urlencoded'], "oauth_token={$tmpToken}&oauth_token_secret={$tmpTokenSecret}&oauth_callback_confirmed=1"),
            new Response(200, ['Content-Type' => 'application/x-www-form-urlencoded'], "oauth_token={$token}&oauth_token_secret={$tokenSecret}"),
            new Response(200, ['Content-Type' => 'text/plain'], 'SUCCESS'),
        ]);

        $container = [];
        $history = Middleware::history($container);

        $stack = HandlerStack::create($mock);
        $stack->push($history);

        $client = new Client(['handler' => $stack]);
        $oauth  = new Consumer($client);

        // request token
        $url      = new Url('http://127.0.0.1/requestToken');
        $response = $oauth->requestToken($url, self::CONSUMER_KEY, self::CONSUMER_SECRET);

        $this->assertInstanceOf(Data\Response::class, $response);
        $this->assertEquals(self::TMP_TOKEN, $response->getToken());
        $this->assertEquals(self::TMP_TOKEN_SECRET, $response->getTokenSecret());

        // if we have optained temporary credentials we can redirect the user
        // to grant access to the credentials
        // $oauth->userAuthorization($url, array('oauth_token' => $response->getToken()))

        // if the user gets redirected back we can exchange the temporary
        // credentials to an access token we also get an verifier as GET
        // parameter
        $url      = new Url('http://127.0.0.1/accessToken');
        $response = $oauth->accessToken($url, self::CONSUMER_KEY, self::CONSUMER_SECRET, self::TMP_TOKEN, self::TMP_TOKEN, self::VERIFIER);

        $this->assertInstanceOf(Data\Response::class, $response);
        $this->assertEquals(self::TOKEN, $response->getToken());
        $this->assertEquals(self::TOKEN_SECRET, $response->getTokenSecret());

        // now we can make an request to the protected api
        $url      = new Url('http://127.0.0.1/api');
        $auth     = $oauth->getAuthorizationHeader($url, self::CONSUMER_KEY, self::CONSUMER_SECRET, self::TOKEN, self::TOKEN_SECRET, 'HMAC-SHA1', 'GET');
        $request  = new GetRequest($url, array('Authorization' => $auth));
        $response = $client->request($request);

        $this->assertEquals(200, $response->getStatusCode());
        $this->assertEquals('SUCCESS', (string) $response->getBody());

        $this->assertEquals(3, count($container));
        $transaction = array_shift($container);

        $header = $transaction['request']->getHeaderLine('Authorization');
        $auth   = Authentication::decodeParameters($header);

        $this->assertEquals(self::CONSUMER_KEY, $auth['oauth_consumer_key']);
        $this->assertEquals('HMAC-SHA1', $auth['oauth_signature_method']);
        $this->assertTrue(isset($auth['oauth_timestamp']));
        $this->assertTrue(isset($auth['oauth_nonce']));
        $this->assertEquals('1.0', $auth['oauth_version']);
        $this->assertEquals('oob', $auth['oauth_callback']);
        $this->assertTrue(isset($auth['oauth_signature']));
        
        $transaction = array_shift($container);

        $header = $transaction['request']->getHeaderLine('Authorization');
        $auth   = Authentication::decodeParameters($header);

        $this->assertEquals(self::CONSUMER_KEY, $auth['oauth_consumer_key']);
        $this->assertEquals(self::TMP_TOKEN, $auth['oauth_token']);
        $this->assertEquals('HMAC-SHA1', $auth['oauth_signature_method']);
        $this->assertTrue(isset($auth['oauth_timestamp']));
        $this->assertTrue(isset($auth['oauth_nonce']));
        $this->assertEquals('1.0', $auth['oauth_version']);
        $this->assertEquals(self::VERIFIER, $auth['oauth_verifier']);
        $this->assertTrue(isset($auth['oauth_signature']));

        $transaction = array_shift($container);

        $header = $transaction['request']->getHeaderLine('Authorization');
        $auth   = Authentication::decodeParameters($header);

        $this->assertEquals(self::CONSUMER_KEY, $auth['oauth_consumer_key']);
        $this->assertEquals(self::TOKEN, $auth['oauth_token']);
        $this->assertEquals('HMAC-SHA1', $auth['oauth_signature_method']);
        $this->assertTrue(isset($auth['oauth_timestamp']));
        $this->assertTrue(isset($auth['oauth_nonce']));
        $this->assertEquals('1.0', $auth['oauth_version']);
        $this->assertTrue(isset($auth['oauth_signature']));
    }

    public function testOAuthBuildAuthString()
    {
        $data = Consumer::buildAuthString(array('fo o' => 'b~ar'));

        $this->assertEquals('fo%20o="b~ar"', $data);
    }

    public function testOAuthGetNormalizedUrl()
    {
        $url = new Url('HTTP://Example.com:80/resource?id=123');

        $this->assertEquals('http://example.com/resource', Consumer::getNormalizedUrl($url));


        $url = new Url('http://localhost:8888/amun/public/index.php/api/auth/request');

        $this->assertEquals('http://localhost:8888/amun/public/index.php/api/auth/request', Consumer::getNormalizedUrl($url));
    }

    /**
     * Tests the getNormalizedParameters function by using the values from the
     * RFC. The one "a3" value is commented because we cant have two keys with
     * the same name in an array to make it work we have to change the
     * datastructure but by now no problems occured with this issue
     *
     * @see http://tools.ietf.org/html/rfc5849#section-3.4.2
     * @see http://wiki.oauth.net/w/page/12238556/TestCases
     */
    public function testOAuthGetNormalizedParameters()
    {
        $params = Consumer::getNormalizedParameters(array(

            'b5' => '=%3D',
            //'a3' => 'a',
            'c@' => '',
            'a2' => 'r b',
            'oauth_consumer_key' => '9djdj82h48djs9d2',
            'oauth_token' => 'kkk9d7dh3k39sjv7',
            'oauth_signature_method' => 'HMAC-SHA1',
            'oauth_timestamp' => '137131201',
            'oauth_nonce' => '7d8f3e4a',
            'c2' => '',
            'a3' => '2 q',

        ));

        $expect = 'a2=r%20b&a3=2%20q&b5=%3D%253D&c%40=&c2=&oauth_consumer_key=9djdj82h48djs9d2&oauth_nonce=7d8f3e4a&oauth_signature_method=HMAC-SHA1&oauth_timestamp=137131201&oauth_token=kkk9d7dh3k39sjv7';

        $this->assertEquals($expect, $params);


        $params = array(
            'name='       => array('name' => ''),
            'a=b'         => array('a' => 'b'),
            'a=b&c=d'     => array('a' => 'b', 'c' => 'd'),
            'a=x%2By'     => array('a' => 'x+y'),
            'x=a&x%21y=a' => array('x!y' => 'a', 'x' => 'a'),
        );

        foreach ($params as $expect => $param) {
            $this->assertEquals($expect, Consumer::getNormalizedParameters($param));
        }
    }

    /**
     * Tests url encoding
     *
     * @see http://wiki.oauth.net/w/page/12238556/TestCases
     */
    public function testParameterEncoding()
    {
        $values = array(
            'abcABC123' => 'abcABC123',
            '-._~'      => '-._~',
            '%'         => '%25',
            '+'         => '%2B',
            '&=*'       => '%26%3D%2A',
            "\x0A"      => '%0A',
            "\x20"      => '%20',
            //"\x80"      => '%C2%80',
        );

        foreach ($values as $k => $v) {
            $this->assertEquals($v, Consumer::urlEncode($k));
        }
    }
}


tests/Signature/RSASHA1Test.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth\Tests\Signature;

use PHPUnit\Framework\TestCase;

/**
 * RSASHA1Test
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 */
class RSASHA1Test extends TestCase
{
    public function testNotImplemented()
    {
    }
}


tests/Signature/HMACSHA1Test.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth\Tests\Signature;

use PHPUnit\Framework\TestCase;
use PSX\Oauth\Signature\HMACSHA1;

/**
 * HMACSHA1Test
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 */
class HMACSHA1Test extends TestCase
{
    public function testSignature()
    {
        $signature = new HMACSHA1();

        $this->assertEquals('yUsI5jibqNBZ9W+XPOJaeyrwluM=', $signature->build('', 'djr9rjt0jd78jf88', 'jjd999tj88uiths3'));
        $this->assertEquals('DiK4c9nwDwPEAQPV1XL153JMKvc=', $signature->build('', 'djr9rjt0jd78jf88', 'jjd99$tj88uiths3'));
        $this->assertEquals('QdIl7ldiQOL0oFLNUENqw/mw3ls=', $signature->build('', 'djr9rjt0jd78jf88'));

        $this->assertEquals('egQqG5AJep5sJ7anhXju1unge2I=', $signature->build('bs', 'cs', ''));
        $this->assertEquals('VZVjXceV7JgPq/dOTnNmEfO0Fv8=', $signature->build('bs', 'cs', 'ts'));
        $this->assertEquals('tR3+Ty81lMeYAr/Fid0kMTYa/WM=', $signature->build('GET&http%3A%2F%2Fphotos.example.net%2Fphotos&file%3Dvacation.jpg%26oauth_consumer_key%3Ddpf43f3p2l4k3l03%26oauth_nonce%3Dkllo9940pd9333jh%26oauth_signature_method%3DHMAC-SHA1%26oauth_timestamp%3D1191242096%26oauth_token%3Dnnch734d00sl2jdk%26oauth_version%3D1.0%26size%3Doriginal', 'kd94hf93k423kf44', 'pfkkdhi9sl3r4s00'));
    }
}


src/Data/Request.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth\Data;

use PSX\Data\InvalidDataException;
use PSX\Record\Record;
use PSX\Validate;

/**
 * Request
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 */
class Request extends Record
{
    public function setConsumerKey($consumerKey)
    {
        $this->setProperty('oauth_consumer_key', $consumerKey);
    }
    
    public function getConsumerKey()
    {
        return $this->getProperty('oauth_consumer_key');
    }

    public function setToken($token)
    {
        $this->setProperty('oauth_token', $token);
    }
    
    public function getToken()
    {
        return $this->getProperty('oauth_token');
    }

    public function setSignatureMethod($signatureMethod)
    {
        switch ($signatureMethod) {
            case 'HMAC-SHA1':
            case 'RSA-SHA1':
            case 'PLAINTEXT':
                $this->setProperty('oauth_signature_method', $signatureMethod);
                break;

            default:
                throw new InvalidDataException('Invalid signature method');
                break;
        }
    }

    public function getSignatureMethod()
    {
        return $this->getProperty('oauth_signature_method');
    }

    public function setSignature($signature)
    {
        $this->setProperty('oauth_signature', $signature);
    }

    public function getSignature()
    {
        return $this->getProperty('oauth_signature');
    }

    /**
     * @param integer $timestamp
     */
    public function setTimestamp($timestamp)
    {
        if (is_numeric($timestamp) && strlen($timestamp) == 10) {
            $this->setProperty('oauth_timestamp', $timestamp);
        } else {
            throw new InvalidDataException('Invalid timestamp format');
        }
    }

    public function getTimestamp()
    {
        return $this->getProperty('oauth_timestamp');
    }

    public function setNonce($nonce)
    {
        $this->setProperty('oauth_nonce', $nonce);
    }

    public function getNonce()
    {
        return $this->getProperty('oauth_nonce');
    }

    public function setCallback($callback)
    {
        if ($callback == 'oob') {
            // callback was set "out of bound" ... we get the url later from the
            // consumer object
            $this->setProperty('oauth_callback', 'oob');
        } elseif (strlen($callback) >= 7 && strlen($callback) <= 256 && filter_var($callback, FILTER_VALIDATE_URL, FILTER_FLAG_SCHEME_REQUIRED)) {
            $this->setProperty('oauth_callback', $callback);
        } else {
            throw new InvalidDataException('Invalid callback format');
        }
    }

    public function getCallback()
    {
        return $this->getProperty('oauth_callback');
    }

    public function setVersion($version)
    {
        $this->setProperty('oauth_version', $version);
    }

    public function getVersion()
    {
        return $this->getProperty('oauth_version');
    }

    public function setVerifier($verifier)
    {
        if (strlen($verifier) >= 16 && strlen($verifier) <= 512) {
            $this->setProperty('oauth_verifier', $verifier);
        } else {
            throw new InvalidDataException('Invalid verifier format');
        }
    }

    public function getVerifier()
    {
        return $this->getProperty('oauth_verifier');
    }
}


src/ResponseImporter.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth;

use InvalidArgumentException;
use PSX\Oauth\Data\Response;

/**
 * ResponseImporter
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 */
class ResponseImporter
{
    public function import(Response $record, $data)
    {
        if (!is_array($data) && !$data instanceof \stdClass) {
            throw new InvalidArgumentException('Received invalid data');
        }

        foreach ($data as $key => $value) {
            switch ($key) {
                case 'oauth_token':
                    $record->setToken($value);
                    break;

                case 'oauth_token_secret':
                    $record->setTokenSecret($value);
                    break;

                default:
                    $record->addParam($key, $value);
                    break;
            }
        }

        return $record;
    }
}


src/SignatureAbstract.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth;

use PSX\Oauth\Consumer;

/**
 * SignatureAbstract
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 */
abstract class SignatureAbstract
{
    /**
     * Creates a signature from the base string with the consumer secret
     * as key. If the token secret is avialable it is append to the key.
     * Returns the base64 encoded signature
     *
     * @see http://oauth.net/core/1.0a#rfc.section.9
     * @param string $baseString
     * @param string $consumerSecret
     * @param string $tokenSecret
     * @return string
     */
    abstract public function build($baseString, $consumerSecret, $tokenSecret = '');

    /**
     * Compares whether the $signature is valid by creating a new signature
     * and comparing them with $signature
     *
     * @param string $baseString
     * @param string $consumerSecret
     * @param string $tokenSecret
     * @param string $signature
     * @return boolean
     */
    public function verify($baseString, $consumerSecret, $tokenSecret = '', $signature)
    {
        $lft = Consumer::urlDecode($signature);
        $rgt = Consumer::urlDecode($this->build($baseString, $consumerSecret, $tokenSecret));

        return strcasecmp($lft, $rgt) == 0;
    }
}


src/Consumer.php
<?php
/*
 * PSX is a open source PHP framework to develop RESTful APIs.
 * For the current version and informations visit <http://phpsx.org>
 *
 * Copyright 2010-2018 Christoph Kappestein <christoph.kappestein@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace PSX\Oauth;

use PSX\Data\Reader\Form;
use PSX\Framework\Base;
use PSX\Http\Client;
use PSX\Http\PostRequest;
use PSX\Oauth\Data\Response;
use PSX\Oauth\ResponseImporter;
use PSX\Oauth\Signature;
use PSX\Uri\Url;
use RuntimeException;

/**
 * This is a consumer implementation of OAuth Core 1.0
 *
 * @author  Christoph Kappestein <christoph.kappestein@gmail.com>
 * @author  Andy Smith <http://term.ie>
 * @license http://www.apache.org/licenses/LICENSE-2.0
 * @link    http://phpsx.org
 * @see     http://tools.ietf.org/html/rfc5849
 */
class Consumer
{
    /**
     * @var \PSX\Http\Client
     */
    protected $client;

    /**
     * @var \PSX\Data\ReaderInterface
     */
    protected $reader;

    /**
     * @var \PSX\Oauth\ResponseImporter
     */
    protected $importer;

    public function __construct(Client $client)
    {
        $this->client   = $client;
        $this->reader   = new Form();
        $this->importer = new ResponseImporter();
    }

    /**
     * Requests a new "request token" from the $url using the consumer key and
     * secret. The $url must be valid request token endpoint. Returns an array
     * with all key values pairs from the response i.e.
     * <code>
     * $response = $oauth->requestToken(...);
     *
     * $token       = $response->getToken();
     * $tokenSecret = $response->getTokenSecret();
     * </code>
     *
     * @see http://tools.ietf.org/html/rfc5849#section-2.1
     * @param \PSX\Uri\Url $url
     * @param string $consumerKey
     * @param string $consumerSecret
     * @param string $method
     * @param string $callback
     * @return \PSX\Oauth\Data\Response
     */
    public function requestToken(Url $url, $consumerKey, $consumerSecret, $method = 'HMAC-SHA1', $callback = null)
    {
        $values = array(
            'oauth_consumer_key'     => $consumerKey,
            'oauth_signature_method' => $method,
            'oauth_timestamp'        => self::getTimestamp(),
            'oauth_nonce'            => self::getNonce(),
            'oauth_version'          => self::getVersion(),
        );

        // if we have an callback add them to the request
        if (!empty($callback)) {
            $values['oauth_callback'] = $callback;
        } else {
            $values['oauth_callback'] = 'oob';
        }

        // build the base string
        $requestMethod = 'POST';
        $params        = array_merge($values, $url->getParameters());
        $baseString    = self::buildBasestring($requestMethod, $url, $params);

        // get the signature object
        $signature = self::getSignature($method);

        // generate the signature
        $values['oauth_signature'] = $signature->build($baseString, $consumerSecret);

        // request unauthorized token
        $request  = new PostRequest($url, array(
            'Authorization' => 'OAuth realm="psx", ' . self::buildAuthString($values),
            'User-Agent'    => __CLASS__,
        ));
        $response = $this->client->request($request);

        // parse the response
        return $this->importer->import(new Response(), $this->reader->read($response->getBody()));
    }

    /**
     * Exchange an request token with an access token. We receive the "token"
     * and "verifier" from the service provider wich redirects the user to the
     * callback in this redirect are the $token and $verifier. Returns an access
     * token and secret i.e.
     * <code>
     * $response = $oauth->accessToken(...);
     *
     * $token       = $response->getToken();
     * $tokenSecret = $response->getTokenSecret();
     * </code>
     *
     * @see http://tools.ietf.org/html/rfc5849#section-2.3
     * @param \PSX\Uri\Url $url
     * @param string $consumerKey
     * @param string $consumerSecret
     * @param string $token
     * @param string $tokenSecret
     * @param string $verifier
     * @param string $method
     * @return \PSX\Oauth\Data\Response
     */
    public function accessToken(Url $url, $consumerKey, $consumerSecret, $token, $tokenSecret, $verifier, $method = 'HMAC-SHA1')
    {
        $values = array(
            'oauth_consumer_key'     => $consumerKey,
            'oauth_token'            => $token,
            'oauth_signature_method' => $method,
            'oauth_timestamp'        => self::getTimestamp(),
            'oauth_nonce'            => self::getNonce(),
            'oauth_version'          => self::getVersion(),
            'oauth_verifier'         => $verifier,
        );

        // build the base string
        $requestMethod = 'POST';
        $params        = array_merge($values, $url->getParameters());
        $baseString    = self::buildBasestring($requestMethod, $url, $params);

        // get the signature object
        $signature = self::getSignature($method);

        // generate the signature
        $values['oauth_signature'] = $signature->build($baseString, $consumerSecret, $tokenSecret);

        // request access token
        $request  = new PostRequest($url, array(
            'Authorization' => 'OAuth realm="psx", ' . self::buildAuthString($values),
            'User-Agent'    => __CLASS__,
        ));
        $response = $this->client->request($request);

        // parse the response
        return $this->importer->import(new Response(), $this->reader->read($response->getBody()));
    }

    /**
     * If you have established a token and token secret you can use this method
     * to get the authorization header. You can add the header to an http
     * request to make an valid oauth request i.e.
     * <code>
     * $header = array(
     * 	'Authorization: ' . $oauth->getAuthorizationHeader(...),
     * );
     * </code>
     *
     * @param \PSX\Uri\Url $url
     * @param string $consumerKey
     * @param string $consumerSecret
     * @param string $token
     * @param string $tokenSecret
     * @param string $method
     * @param string $requestMethod
     * @param array $post
     * @return string
     */
    public function getAuthorizationHeader(Url $url, $consumerKey, $consumerSecret, $token, $tokenSecret, $method = 'HMAC-SHA1', $requestMethod = 'GET', array $post = array())
    {
        $values = array(
            'oauth_consumer_key'     => $consumerKey,
            'oauth_token'            => $token,
            'oauth_signature_method' => $method,
            'oauth_timestamp'        => self::getTimestamp(),
            'oauth_nonce'            => self::getNonce(),
            'oauth_version'          => self::getVersion(),
        );

        // build the base string
        $params = array_merge($values, $url->getParameters());

        if ($requestMethod == 'POST' && !empty($post)) {
            $params = array_merge($params, $post);
        }

        $baseString = self::buildBasestring($requestMethod, $url, $params);

        // get the signature object
        $signature = self::getSignature($method);

        // generate the signature
        $values['oauth_signature'] = $signature->build($baseString, $consumerSecret, $tokenSecret);

        // build request
        $authorizationHeader = 'OAuth realm="psx", ' . self::buildAuthString($values);

        return $authorizationHeader;
    }

    /**
     * Returns the signature object based on the $method throws an exception if
     * the method is not supported
     *
     * @param string $method
     * @return \PSX\Oauth\SignatureAbstract
     */
    public static function getSignature($method)
    {
        switch ($method) {
            case 'HMAC-SHA1':
                return new Signature\HMACSHA1();
                break;

            case 'RSA-SHA1':
                return new Signature\RSASHA1();
                break;

            case 'PLAINTEXT':
                return new Signature\PLAINTEXT();
                break;

            default:
                throw new RuntimeException('Invalid signature method');
                break;
        }
    }

    /**
     * Build the string that we use in the authentication header
     *
     * @param array $data
     * @return string
     */
    public static function buildAuthString(array $data)
    {
        $str = array();

        foreach ($data as $k => $v) {
            $str[] = self::urlEncode($k) . '="' . self::urlEncode($v) . '"';
        }

        return implode(', ', $str);
    }

    /**
     * Builds the basestring for the signature.
     *
     * @see http://tools.ietf.org/html/rfc5849#section-3.4.1
     * @param string $method
     * @param \PSX\Uri\Url $url
     * @param array $data
     * @return string
     */
    public static function buildBasestring($method, Url $url, array $data)
    {
        $base = array();
        $base[] = self::urlEncode(self::getNormalizedMethod($method));
        $base[] = self::urlEncode(self::getNormalizedUrl($url));
        $base[] = self::urlEncode(self::getNormalizedParameters($data));

        return implode('&', $base);
    }

    /**
     * Returns the method in uppercase
     *
     * @param string $method
     * @return string
     */
    public static function getNormalizedMethod($method)
    {
        return strtoupper($method);
    }

    /**
     * Normalize the url like defined in
     *
     * @see http://tools.ietf.org/html/rfc5849#section-3.4.1.2
     * @param \PSX\Uri\Url $url
     * @return string
     */
    public static function getNormalizedUrl(Url $url)
    {
        $scheme = $url->getScheme();
        $host   = $url->getHost();
        $port   = $url->getPort();
        $path   = $url->getPath();

        // no port for 80 (http) and 443 (https)
        if ((($port == 80 || empty($port)) && strcasecmp($scheme, 'http') == 0) || (($port == 443 || empty($port)) && strcasecmp($scheme, 'https') == 0)) {
            $normalizedUrl = $scheme . '://' . $host . $path;
        } else {
            if (!empty($port)) {
                $normalizedUrl = $scheme . '://' . $host . ':' . $port . $path;
            } else {
                throw new RuntimeException('No port specified');
            }
        }

        return strtolower($normalizedUrl);
    }

    /**
     * Returns the parameters that we need to create the basestring
     *
     * @param array $data
     * @return string
     */
    public static function getNormalizedParameters(array $data)
    {
        $params = array();

        $keys   = array_map('PSX\Oauth\Consumer::urlEncode', array_keys($data));
        $values = array_map('PSX\Oauth\Consumer::urlEncode', array_values($data));
        $data   = array_combine($keys, $values);


        uksort($data, 'strnatcmp');


        foreach ($data as $k => $v) {
            if ($k != 'oauth_signature') {
                $params[] = $k . '=' . $v;
            }
        }

        return implode('&', $params);
    }

    /**
     * Encode values RFC3986
     *
     * @see http://tools.ietf.org/html/rfc5849#section-3.6
     * @param string $data
     * @return string
     */
    public static function urlEncode($data)
    {
        return str_replace('%7E', '~', rawurlencode($data));
    }

    /**
     * Decode values RFC3986
     *
     * @param string $data
     * @return string
     */
    public static function urlDecode($data)
    {
        return rawurldecode($data);
    }

    /**
     * Returns the current timestamp used in a request
     *
     * @return integer
     */
    public static function getTimestamp()
    {
        return time();
    }

    /**
     * Returns the nonce used in a request
     *
     * @return string
     */
    public static function getNonce()
    {
        return substr(md5(uniqid(mt_rand(), true)), 0, 16);
    }

    /**
     * Returns the current version use in a request
     *
     * @return string
     */
    public static function getVersion()
    {
        return '1.0';
    }

    /**
     * This method returns an array with the support methods to sign a request
     *
     * @return array
     */
    public static function getSupportedMethods()
    {
        return array('HMAC-SHA1', 'PLAINTEXT');
    }
}
